File: kotlin/lesson 2/ArrayStack.kt
class ArrayStack<T>() : Stack<T> {
    var items = Array<Any?>(size = INITIAL_CAPACITY, { null }) as Array<T>
    var count = 0
    
    override fun push(value: T) {
        if (count == items.size) resize(items.size * 2)
        items[count] = value
        count++
    }
    
    override fun pop(): T {
        count--
        val result = items[count]
        if (count > 0 && count == items.size / 4) resize(items.size / 2)
        return result
    }	
    
    inline fun resize(size: Int) {
		val copy = Array<Any?>(size, { null }) as Array<T>
        for (i in 0..this.count) {
            copy[i] = items[i]
        }
        items = copy
    }
    
    companion object {
        const val INITIAL_CAPACITY = 2
	}
}

File: csharp/lesson 2/ArrayStack.cs
public class ArrayStack<T> : Stack<T>
{
    private const int INITIAL_CAPACITY = 2;
    private T[] items = new T[INITIAL_CAPACITY];
    private int count;

    public void Push(T value)
    {        
        if (count == items.Length) Resize(items.Length * 2);
        items[count] = value;
        count++;
    }

    public T Pop()
    {
        count--;
        T result = items[count];
        items[count] = default(T);
        if (count > 0 && count == items.Length / 4) Resize(items.Length / 2);
        return result;
    }

    private void Resize(int size) 
    {
        T[] copy = new T[size];
        for (int i = 0; i < this.count; i++)
        {
            copy[i] = items[i];
        }
        items = copy;
    }
}